### ISSUE TITLE
Erratic results of find in page 

### ISSUE THREAD

**Issue by csadilek** [mozilla-mobile/android-components#4704]
STR:
- Load a page (cnn.com works well for reproducing)
- Step 1: Type 'a', then type 'n' 
- Step 2: Hit ^ to display next result a few times in a row (quickly)

Expected after Step 1:
We'd expect the result counts of 'a' then the result counts of 'n'. 

Actual after Step 1:
Sometimes after typing 'n' we still show the result counts of 'a'.

Expected after Step 2:
Iterating through the results causes the counter to increment/decrement and the total results to stay stable.

Actual after Step 2:
Sometimes we display 0/0, sometimes the total results change.

The problem of the incorrect result count in step 2, I can also reproduce in Fennec, but it definitely works much better there.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-3957)

**Comment by csadilek** [comment]
Tested this with latest R-B and Fenix release builds.

**Comment by csadilek** [comment]
This looks like it was also a problem before our recent refactoring. When calling `findAll` in quick succession, there's no guarantee that the results come back in order. So, while we always display the last result, it doesn't mean that the last result is from the last search (it could be from a previous input character).

It seems we need a similar mechanism to our awesomebar where we either cancel previous searches or discard stale results. Wdyt @pocmo @Amejia481 ?

[1] https://github.com/mozilla-mobile/android-components/blob/master/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt#L219
[2] https://dxr.mozilla.org/mozilla-central/source/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/SessionFinder.java?q=SessionFinder.java&redirect_type=direct#84

**Comment by pocmo** [comment]
Yeah, that makes totally sense.

If we'd already use the Store from the engine I'd say we could store the search string in the state and then in the reducer only update when the result is for the same search string.

Maybe we could emulate this? We already pass the text to `EngineObserver.onFind()` but we only use it to clear the list. If we keep it then we can use that once `EngineObserver.onFindResult()` gets called (onFindResult() would need to tell what text it is for though!).

**Comment by csadilek** [comment]
@pocmo after some more testing, I think this is a GV bug. The `GeckoSession.FinderResult`s contain the search string and I see those results come back in the right order (with the right search string), but with inconsistent values. I also see interleaved 0/0 results come back on find next (when called in quick succession) which are incorrect too.

So, I've verified that even when things happen in the right order the results we get back can be incorrect. I'll file an issue.

**Comment by csadilek** [comment]
GV bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1587856

**Comment by pocmo** [comment]
RB: https://github.com/mozilla-mobile/reference-browser/issues/1020

**Comment by pocmo** [comment]
Focus: https://github.com/mozilla-mobile/focus-android/issues/5181

**Comment by csadilek** [comment]
Closing in favour of https://bugzilla.mozilla.org/show_bug.cgi?id=1587856.
